### PR TITLE
[FLINK-32414][connectors/common] Don't emitLatestWatermark when lastEmittedWatermark is UNINITIALIZED

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

Watermark alignment will cause flink jobs to hang forever when any source subtask has no SourceSplit. You can get more background from FLINK-32414.

## Brief change log

Don't emitLatestWatermark when lastEmittedWatermark is UNINITIALIZED

## Verifying this change

This change added tests and can be verified as follows:

  - *Added SourceOperatorAlignmentTest#testWatermarkAlignmentWithoutSplit*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers:  no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper:  no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no

### CODE REVIEW

**Comment by 1996fanrui** [comment]
Hi @pnowojski  @LoveHeat @RocMarshal  , would you mind help take a look in your free time? thanks~

I have described the detailed background in FLINK-32414.

**Comment by 1996fanrui** [comment]
Hi @pnowojski  @dawidwys  @LoveHeat @RocMarshal  , would you mind help take a look in your free time? thanks~

This bug is watermark alignment will cause flink jobs to hang forever when any source subtask has no SourceSplit, I have described the bug in detail in FLINK-32414.

**Review by RocMarshal — COMMENTED** [review]
Hi, @1996fanrui Thank you for the contribution.  👍 
I left a few of comments. PTAL in your free time.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java [hunk 1]
```diff
@@ -46,12 +46,7 @@ Licensed to the Apache Software Foundation (ASF) under one
...
 
-import static org.hamcrest.CoreMatchers.is;
-import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.contains;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.assertj.core.api.Assertions.assertThat;
 
 /** Unit test for {@link SourceOperator} watermark alignment. */
```
[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java:51]: Thanks for the review. Good catch! updated.

**Review by RocMarshal — APPROVED** [review]
Thanks for the update.
LGTM +1 (non-binding) now.

**Review by mas-chen — APPROVED** [review]
LGTM! +1

**Comment by 1996fanrui** [comment]
Thanks a lot @RocMarshal  and @mas-chen 's review, merging~

**Pull request merged**